core-bundle/src/Security/Voter/BackendAccessVoter.php

```diff
@@ -121,12 +121,21 @@ private function hasAccess(mixed $subject, string $field, BackendUser $user): bo
 
         // Additionally check the child pages of the mounted pages
         if ('pagemounts' === $field) {
-            if (!isset($this->pagemountsCache[$user->id])) {
+            $subject = array_flip($subject);
+            $matches = array_intersect_key($this->pagemountsCache[$user->id] ?? [], $subject);
+
+            if (\count($matches) !== \count($subject)) {
                 $database = $this->framework->createInstance(Database::class);
-                $this->pagemountsCache[$user->id] = $database->getChildRecords($user->pagemounts, 'tl_page');
+
+                $this->pagemountsCache[$user->id] = array_replace(
+                    array_fill_keys($database->query('SELECT id FROM tl_page')->fetchEach('id'), false),
```

Comment:
This fills the cache with potentially thousands of IDs. How can that be more effective than no cache at all?

Reply:
because this permission check is performed **for almost every operation when rendering the tree** 😉

Reply:
We could optimize this with our preloadHints on the action. The `DC_Table` could hint the `visibleIds` so you could preload only those for example. But that'd be for another PR. But caching thousands of IDs will use very little memory - the problem is not the memory, it's fetching those from the DB.

Reply:
> The `DC_Table` could hint the `visibleIds` so you could preload only those for example

That sounds more reasonable. There are Contao installations with 50,000 pages and more and it certainly does not make sense to load all those IDs if only a few of the pages are visible.

Reply:
As discussed in the Contao call, we should revert the changes and keep line 126 but if a page ID is not in the cache, we should rebuild the cache before finally returning `true` or `false`.

Reply:
As discussed in the Contao call, we should revert the changes and keep line 126 but if a page ID is not in the cache, we should rebuild the cache and look again before finally returning `true` or `false`.